message_ix/models.py

```diff
@@ -378,21 +419,44 @@ class MESSAGE(GAMSModel):
 
     name = "MESSAGE"
 
+    @staticmethod
+    def enforce(scenario):
+        """Enforce data consistency in `scenario`."""
+        # Check masks ("mapping sets") that indicate which elements of corresponding
+        # parameters are active/non-zero. Note that there are other masks currently
+        # handled in JDBCBackend. For the moment, this code does not backstop that
+        # behaviour.
+        # TODO Extend to handle all masks, e.g. for new backends.
+        for par_name in ("bound_emission_pool", "emission_sink_rate"):
+            # Name of the corresponding set
+            set_name = f"is_{par_name}"
+
+            # Existing and expected contents
+            existing = scenario.set(set_name)
+            expected = scenario.par(par_name).drop(columns=["value", "unit"])
+
+            if existing.equals(expected):
+                continue  # Contents are as expected; do nothing
+
+            # Not consistent; empty and then re-populate the set
+            with scenario.transact(f"Enforce consistency of {set_name} and {par_name}"):
+                scenario.remove_set(set_name, existing)
```

Comment:
@OFR-IIASA this line is supposed to entirely clear the set before it is populated by the next line. You could perhaps check whether this is having the intended effect.

Reply:
@khaeru I have added a two tests to see if this feature works correctly. 
- Test 1. adds the parameter `bound_emission_pool` for 2 periods, solves and checks `is_bound_emission_pool`.
   The bound is then extended across another period, solved and checks `is_bound_emission_pool`
- Test 2. adds the parameter `bound_emission_pool` for 3 periods, solves and checks `is_bound_emission_pool`.
   The `bound_emission_pool` is removed.
   The `bound_emission_pool` is then re-added, but only for two periods, solved and checks `is_bound_emission_pool`.
Both the tests work and confirm the working of the code above.

Reply:
@khaeru The above code has tests, but for the functionality for which the code, in my understanding was designed. The design was to ensure that the "is_*" is consistent when the corresponding parameters is being _modified_; the code does not cover the _removal_ of the corresponding parameter. Further, this code is only executed for `model="MESSAGE"`, an not when passing the argument `model="MESSAGE-MACRO"` as part of the solve command.

Reply:
> The design was to ensure that the "is_*" is consistent when the corresponding parameters is being modified; the code does not cover the removal of the corresponding parameter. The removal is test in a test test_bound_emission_pool_removal(), which fails.

Thanks for directing to this specific test. To be clear:
- The added `enforce()` method is called _at the moment the model is run/solved_.
- In the final lines of that test, there is a call to `.remove_par("bound_emission_pool", …)`, then a call to `.commit()` but—unlike the other tests—there is *no call to `.solve()`* before `.set("is_bound_emission_pool")` is checked.
- So `enforce()` is never invoked, and the result is as expected.
- I suspect that if you called `.solve()` before checking the set, then the test would pass.

More detail: a possible choices would be to have `enforce()` run (a) on every call to `.commit()`, or even (b) on every call to `.add_par()`. The current choice is deliberate: it mirrors the behaviour of the Java code underlying JDBCBackend, which only creates these mapping/etc. sets at the moment the data is written out to GDX. That Java code must someday be ported into message_ix. `enforce()` is now the destination for that code and runs at the same point.

I thought this was better than introducing 1+ additional points at which data is massaged automatically; it would be complex and confusing additional work to think through and plan for possible interactions of these automatic changes. We unfortunately do not have the bandwidth for that at the moment.

> Further, this code is only executed for model="MESSAGE", not when passing the argument model="MESSAGE-MACRO" as part of the solve command.

This I can check, and will do after you try adjusting the test.

Reply:
I finally got around to sorting out the tests. All tests pass; now the issue would be to ensure that we have this working when the calling `solve(model="MESSAGE-MACRO")`. Unfortunately I dont think we have a simple test-case which can be adapted and run with macro.

Reply:
I can try to look at this next week, after we've wrapped up the current release.